File: app/internal/reporting/analysis.go
package reporting

import (
	"fmt"
	"strings"
	"time"

	"heritage-mgmt/internal/llm"
)

// LLMClient 是 reporting 消费大模型所需的端口（仅依赖 Chat/Configured）。
// *llm.Client 结构化满足该接口，调用方可注入替身做无网络单测。
type LLMClient interface {
	Chat(messages []llm.Message, opts llm.Options) (string, error)
	Configured() bool
}

// GenerateAnalysis 调用大模型基于报告数据生成分析文本。
// 未配置密钥时返回占位说明（无错误），与原行为一致。
func GenerateAnalysis(client LLMClient, rd ReportData, timeout time.Duration) (string, error) {
	if !client.Configured() {
		return "（未配置大模型API密钥，分析报告功能不可用。）", nil
	}
	proj := rd.Project
	cf := proj.CentralFunding
	tp := proj.TotalPaid
	payRate := "—"
	if cf != nil && tp != nil && *cf > 0 {
		payRate = fmt.Sprintf("%.1f%%", *tp / *cf * 100)
	}

	summary := fmt.Sprintf(`工程名称：%s
文物单位：%s  级别：%s
工程类型：%s  当前状态：%s
批复文号：%s
合同工期：%s 至 %s  实际完工：%s
中央指标：%.0f万元  已支付：%.0f万元  支付率：%s
建设单位：%s
施工单位：%s（资质：%s）
设计单位：%s（资质：%s）
监理单位：%s（资质：%s）
文档总数：%d  完整度：%d%%
缺项：%s
资质校验：%s`,
		proj.Name, proj.UnitName, LevelName(rd.UnitLevel),
		proj.Ptype, proj.Status, proj.ApprovalNo,
		proj.SignDate, proj.ContractEnd, proj.CompleteDate,
		fptr(cf), fptr(tp), payRate,
		proj.OwnerUnit,
		proj.ConstructionUnit, proj.ConstructionQual,
		proj.DesignUnit, proj.DesignQual,
		proj.SupervisionUnit, proj.SupervisionQual,
		len(rd.Documents), rd.Completeness,
		strings.Join(rd.MissingRequired, "、"),
		strings.Join(rd.QualWarnings, "；"),
	)

	prompt := `你是文物保护工程档案分析专家。请根据以下工程数据，生成一份专业、正式的工程管理分析报告（约500-800字）。报告应包含：
1. 工程概况评价
2. 资金执行评价（支付率是否合理）
3. 进度评价（工期是否逾期）
4. 参建单位资质评价
5. 档案完整性评价（缺项分析）
6. 综合结论与建议
语气正式客观，不用markdown，直接输出纯文本。

工程数据：
` + summary

	return client.Chat(
		[]llm.Message{{Role: "user", Content: prompt}},
		llm.Options{Temperature: 0.3, MaxTokens: 2000, Timeout: timeout},
	)
}

func fptr(v *float64) float64 {
	if v == nil {
		return 0
	}
	return *v
}

File: desktop/src/core/report/Analysis.cpp
#include "Analysis.h"

#include <QStringList>

namespace heritage::report {

namespace {
QString fptr(const std::optional<double>& v) { return v.has_value() ? QString::number(*v, 'f', 0) : QStringLiteral("0"); }
} // namespace

QString buildAnalysisPrompt(const ReportData& rd) {
    const Project& p = rd.project;
    QString payRate = QStringLiteral("—");
    if (p.centralFunding.has_value() && p.totalPaid.has_value() && *p.centralFunding > 0)
        payRate = QString::number(*p.totalPaid / *p.centralFunding * 100.0, 'f', 1) + QStringLiteral("%");

    const QString summary = QStringLiteral(
        "工程名称：%1\n"
        "文物单位：%2  级别：%3\n"
        "工程类型：%4  当前状态：%5\n"
        "批复文号：%6\n"
        "合同工期：%7 至 %8  实际完工：%9\n"
        "中央指标：%10万元  已支付：%11万元  支付率：%12\n"
        "建设单位：%13\n"
        "施工单位：%14（资质：%15）\n"
        "设计单位：%16（资质：%17）\n"
        "监理单位：%18（资质：%19）\n"
        "文档总数：%20  完整度：%21%\n"
        "缺项：%22\n"
        "资质校验：%23")
        .arg(p.name, p.unitName, levelName(rd.unitLevel), p.ptype, p.status, p.approvalNo,
             p.signDate, p.contractEnd, p.completeDate, fptr(p.centralFunding), fptr(p.totalPaid),
             payRate, p.ownerUnit, p.constructionUnit, p.constructionQual, p.designUnit,
             p.designQual, p.supervisionUnit, p.supervisionQual,
             QString::number(rd.documents.size()), QString::number(rd.completeness),
             rd.missingRequired.isEmpty() ? QStringLiteral("无") : rd.missingRequired.join(QStringLiteral("、")),
             rd.qualWarnings.isEmpty() ? QStringLiteral("无") : rd.qualWarnings.join(QStringLiteral("；")));

    return QStringLiteral(
        "你是文物保护工程档案分析专家。请根据以下工程数据，生成一份专业、正式的工程管理分析报告"
        "（约500-800字）。报告应包含：\n"
        "1. 工程概况评价\n2. 资金执行评价（支付率是否合理）\n3. 进度评价（工期是否逾期）\n"
        "4. 参建单位资质评价\n5. 档案完整性评价（缺项分析）\n6. 综合结论与建议\n"
        "语气正式客观，不用markdown，直接输出纯文本。\n\n工程数据：\n") + summary;
}

QString generateAnalysis(llm::Client& client, const ReportData& rd, int timeoutMs, QString* err) {
    if (err) err->clear();
    if (!client.configured())
        return QStringLiteral("（未配置大模型API密钥，分析报告功能不可用。）");
    llm::Options opts;
    opts.temperature = 0.3;
    opts.maxTokens = 2000;
    opts.timeoutMs = timeoutMs > 0 ? timeoutMs : 90000;
    return client.chat({{QStringLiteral("user"), buildAnalysisPrompt(rd)}}, opts, err);
}

} // namespace heritage::report
